carcara/src/ast/term.rs

```diff
@@ -569,6 +576,68 @@ impl From<SortedVar> for Term {
     }
 }
 
+impl Sort {
+    // Whether this sort can be unified with another. The map argument
+    // will be a substitution of sort variables to sorts
+    pub fn match_with(&self, target: &Sort, map: &mut IndexMap<String, Sort>) -> bool {
+        match (self, target) {
+            (Sort::Var(a), _) => {
+                match map.entry(a.to_string()) {
+                    Entry::Vacant(e) => {
+                        e.insert(target.clone());
+                    }
+                    Entry::Occupied(e) => {
+                        return e.get() == target;
+                    }
+                }
+                true
+            }
+            (Sort::Atom(a, sorts_a), Sort::Atom(b, sorts_b)) => {
+                if a != b {
+                    false
+                } else {
+                    let matching = sorts_a
+                        .iter()
+                        .zip(sorts_b.iter())
+                        .filter(|&(t_a, t_b)| {
+                            let s_a = t_a.as_sort().unwrap();
+                            let s_b = t_b.as_sort().unwrap();
+                            s_a.match_with(s_b, map)
+                        })
+                        .count();
+                    matching == sorts_a.len() && matching == sorts_b.len()
```

Comment:
Here, you can use [`Iterator::all`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all):
```suggestion
                    sorts_a.iter().zip(sorts_b.iter()).all(|(t_a, t_b)| {
                        let s_a = t_a.as_sort().unwrap();
                        let s_b = t_b.as_sort().unwrap();
                        s_a.match_with(s_b, map)
                    })
```

Reply:
Cool!